api/src/indigo_internal.h

```diff
@@ -162,6 +162,7 @@ class DLLEXPORT IndigoObject
         INDIGO_OBJECT_LAST_TYPE // must be the last element in the enum
     };
 
+    int id;

```

Comment:
This change looks attractive, but in fact is discussible. (For example, what would you do, if the stored id mismatches with id key ).
@mkviatkovskii , what do you think?

Reply:
The id is assigned here and only here on object creation. If it changed during the object lifecycle, it would be a bug in many places because the id identifies object for host language (Java.Py etc.) code. So, it can't mismatch or nothing works.

I've introduced the field to walk around architectural inconsistency:  in C++ API, Indigo objects sometimes are referenced by handle (i.e. the id), sometimes by pointer.

int handle => IndigoObject pointer conversion is provided by indigoGetInstance().getObject(int item_handle)
but
IndigoObject pointer => int handle conversion is not provided.

Particularly for me, the problem is here:
static void checkMolecule(**const IndigoObject& item**, int check_types, const std::vector<int>& selected_atoms, const std::vector<int>& selected_bonds,
                          StructureChecker2::CheckResult& result)
{
    checkMolecule(((IndigoObject&)item).getBaseMolecule(), check_types, selected_atoms, selected_bonds, result);
    **check_v3000(item._id,_ result);**
}

static void check_v3000(int mol, StructureChecker2::CheckResult& result)
{
    const char* f = **indigoMolfile(mol);** _<= handle required, no API for IndigoObject&_
    if (f && std::string(f).find("V3000") != -1)
......
}

Reply:
Yep, you wrote exactly what I meant by "looks attractive".
When I talked about mismatch, I meant mismatch at the higher level: between input parameter `id` in calls to functions those search by `int id` and found object's filed `id`. If I'm not mistaken, this input parameter is a key in internal maps, so it exists separately. Finally, having `id` in two places is a redundancy (or indexing).

Reply:
Yes, id is the object's unique permanent key in internal map.
Now object knows its id.
That's all.
What "mismatch" is possible?

int Indigo::addObject(IndigoObject* obj)
{
    OsLocker lock(_objects_lock);
    obj->id = _next_id++;
    _objects.insert(obj->id, obj);
    return obj->id;
}